### PR TITLE
Added ControllerGetVolume API in CSI driver

### PR DESCRIPTION
**Pull request by vishal-shakya-ssir**
This PR has Implementation of ControllerGetVolume API.
A Controller Plugin MUST provide a non-empty volume_condition field in ControllerGetVolumeResponse if it has VOLUME_CONDITION capability.
ControllerGetVolumeResponse should contain current information of a volume if it exists. 
ControllerGetVolume API calls POS VOLUMEINFO API to check volume existence and mount status.

If the volume does not exist or mounted, ControllerGetVolume API will return error message that will subsequently be displayed in K8S events and controller logs.

e.g. 
GRPC call: /csi.v1.Controller/ControllerGetVolume

GRPC response: {"status":{"volume_condition":{Abnormal:true, Message: "9e85e5db-18e4-46b9-912d-6acada8b4266 volume status not found"}},"volume":{"volume_id":"9e85e5db-18e4-46b9-912d-6acada8b4266"}}

### CODE REVIEW

File: src/csi/pkg/pos/controllerserver.go [hunk 1]
```diff
@@ -182,8 +182,47 @@ func (s *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumes
 	return volumeRes, nil
 
 }
+
 func (s *controllerServer) ControllerGetVolume(ctx context.Context, req *csi.ControllerGetVolumeRequest) (*csi.ControllerGetVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	klog.Info("The 'ControllerGetVolume' API function has been successfully populated", req.GetVolumeId())
+	volId := req.GetVolumeId()
+	volume, exists := s.volumesById[volId]
+	message := ""
```
Comment (mjlee34) [inline src/csi/pkg/pos/controllerserver.go:190]: As far as I know, it is better to have the declaration of a variable close to the actual use. Consider to move it to near line 197.

File: src/csi/pkg/pos/dagent.go [hunk 2]
```diff
@@ -70,29 +70,53 @@ type Status struct {
...
+	} else {
+		return model.Response{}, status.Error(codes.Unavailable, response.Result.Status.Description)
+	}
+	return response, nil
+}
+
+func (dagent *DAgent) VolumeInfo(volName string, config map[string]string, mtx2 *sync.Mutex) (model.Response, error) {
+	url := fmt.Sprintf("http://%s:%s/api/ibofos/v1/array/%s/volume/%s", config["provisionerIp"], config["provisionerPort"], config["array"], volName)
+	resp, err := util.CallDAgent(url, nil, "GET", "VolumeInfo", mtx2)
+	if err != nil {
```
Comment (mjlee34) [inline src/csi/pkg/pos/dagent.go:100]: Minor comment: I think the lines are too dense. I suggest to have blank lines between code chunks (a block of related blocks).

File: src/csi/pkg/pos/dagent.go [hunk 3]
```diff
@@ -70,29 +70,53 @@ type Status struct {
...
+		return model.Response{}, status.Error(codes.Unavailable, err.Error())
+	}
+	dec := json.NewDecoder(bytes.NewBuffer(body))
+	dec.UseNumber()
+	response := model.Response{}
+	if err := dec.Decode(&response); err != nil {
+		return model.Response{}, status.Error(codes.Unavailable, err.Error())
+	}
+	if response.Result.Status.Code == 0 {
+		klog.Info("VolumeInfo Success ")
```
Comment (mjlee34) [inline src/csi/pkg/pos/dagent.go:114]: The whitespace at the end of string is intended?

**Review by mjlee34 — CHANGES_REQUESTED** [review]
LGTM, but please check some comments.

File: src/csi/pkg/pos/controllerserver.go [hunk 4]
```diff
@@ -182,8 +182,47 @@ func (s *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumes
...
+		abnormal = true
+		message = volume.name + " volume status not found"
+	} else if response.Result.Status.Code != 0 {
+		abnormal = true
+		message = response.Result.Status.Description
+	} else {
+		data := response.Result.Data.(map[string]interface{})
+		volStatus, keyExist1 := data["status"].(string)
+		if keyExist1 {
+			if volStatus != "Mounted" {
```
[the remark this answers is not in the document]
Comment (vishal-shakya-ssir) [reply src/csi/pkg/pos/controllerserver.go:208]: Sure, Thanks Manoj

File: src/csi/pkg/pos/controllerserver.go (hunk 1 above)
Comment (vishal-shakya-ssir) [reply src/csi/pkg/pos/controllerserver.go:190]: Okay got it, I will do it

File: src/csi/pkg/pos/dagent.go (hunk 3 above)
Comment (vishal-shakya-ssir) [reply src/csi/pkg/pos/dagent.go:114]: I will remove the whitespace

**Review by aswinssir — APPROVED** [review]
LGTM

**Review by mjlee34 — APPROVED** [review]
LGTM

**Pull request merged**